### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by stefan-j** [google/codeworld#177]
This issue exists out of two parts:

- [ ] Large programs are slow when connecting and disconnecting
- [ ] When a Flyout contains too many items, it also starts getting slow..

### The block based editor is slow when snapping, especially for large programs:

This can be seen in program https://code.world/funblocks#PsTt2AwuGXLT2WNb8jLUlUQ (ignoring issue #174) when removing a few blocks and then adding some changes.

I'm speculating this is due to the garbage collection functions in the typed blockly code.

**Comment by stefan-j** [comment]
Regarding large programs, some part of the slow down can possibly be attributed to the disconnect code. Upon disconnected, for each affected block, a block of the same type is created from xml, a new block is instantiated, and the types are copied over.
Disconnecting of blocks is being entirely redone, where only the types are being reset.

I speculated that implementing the block types in Haskell using a callback function would cause a slow down, as the code to construct the proper JS init function is called each time, and only then is the actual initialization function called. However this is not the case. I tested this by re-implementing all the number types in pure JS, however the noticeable delay when opening the Flyout on the numbers category remains the same.

I'm also starting to doubt whether the garbage collection attributes any slowness. I disabled the type variable garbage collection and it also did not have any effect on the Flyout open delay. I'm unsure what it's effect is on the snapping of blocks on large programs.
